File: rust/src/euler035.rs
// https://projecteuler.net/problem=35
//
// The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.
// There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.
// How many circular primes are there below one million?
//
// Answer: 55

fn is_prime(n: u64) -> bool {
    if n < 2 {
        return false;
    }
    if n == 2 || n == 3 {
        return true;
    }
    if n % 2 == 0 || n % 3 == 0 {
        return false;
    }
    let mut i = 5;
    while i * i <= n {
        if n % i == 0 || n % (i + 2) == 0 {
            return false;
        }
        i += 6;
    }
    true
}

pub fn circular_primes() -> u64 {
    let mut count = 0;
    for n in 2..1_000_000 {
        if is_prime(n) {
            let s = format!("{}", n);
            let mut is_circular = true;
            for i in 0..s.len() {
                let rotated: String = s[i..].to_string() + &s[0..i];
                let r: u64 = rotated.parse().unwrap();
                if !is_prime(r) {
                    is_circular = false;
                    break;
                }
            }
            if is_circular {
                count += 1;
            }
        }
    }
    count
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn euler_035() {
        assert_eq!(circular_primes(), 55);
    }
}

File: cpp/src/euler035.cpp
// https://projecteuler.net/problem=35

// The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.

// There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.

// How many circular primes are there below one million?

// Answer: 55

// Authored by: Tim Varley 💘

#include <iostream>
#include <vector>
#include <string>

int circular_primes() {
    const int MAX = 1000000;
    std::vector<bool> is_prime(MAX+1, true);
    is_prime[0] = is_prime[1] = false;
    for(long long i=2; i*i<=MAX; i++) {
        if(is_prime[i]) {
            for(long long j=i*i; j<=MAX; j+=i) {
                is_prime[j] = false;
            }
        }
    }
    int count = 0;
    for(int n=2; n<MAX; n++) {
        if(!is_prime[n]) continue;
        std::string s = std::to_string(n);
        bool all_prime = true;
        for(size_t rot=1; rot<s.size(); rot++) {
            std::string r = s.substr(rot) + s.substr(0,rot);
            int rn = std::stoi(r);
            if(!is_prime[rn]) { all_prime=false; break; }
        }
        if(all_prime) count++;
    }
    return count;
}

#if ! defined UNITTEST_MODE
int main(int argc, char const *argv[]) {
    std::cout << circular_primes() << std::endl;
}
#endif // UNITTEST_MODE